# TPC2: Somador on/off

Crie um programa em Python que tenha o seguinte comportamento:

  * Pretende-se um programa que some todas as sequências de dígitos que encontre num texto;
  * Prepare o programa para ler o texto do canal de entrada: stdin;
  * Sempre que encontrar a string “Off” em qualquer combinação de maiúsculas e minúsculas, esse comportamento é desligado;
  * Sempre que encontrar a string “On” em qualquer combinação de maiúsculas e minúsculas, esse comportamento é novamente ligado;
  * Sempre que encontrar o caráter “=”, o resultado da soma é colocado na saída.


In [2]:
def main(): # incui números negativos e com casas decimais
    
    # ignorar caracteres lidos   ->    texto = texto[1:]   (todos, a partir do índice 1)
    # texto = texto[0::3] -> CUIDADO: ignora também o elemento de índice 3 
    
    somar = 1 # o programa tem o comportamento ligado
    soma_total = 0
    negativo = 0 # 0 -> positivo; 1 -> negativo

    # ler texto do canal de entrada - stdin (CONSTANTEMENTE)
    while 1:
        texto = input("Escreva uma linha: ")
        while texto != "": # final da linha
            
            # Temos de verificar se o próximo número começa por um sinal (negativo)
            if texto[0] == '-': 
                negativo = 1
                texto = texto[1:]

            elif texto[0] == '=': # Apresentar soma total
                print("Soma: " + str(soma_total))
                texto = texto[1:]
            
            # Comportamento desligado (somar = 0) -> Só temos de procurar por 'On'
            elif somar == 0:
                # tem de incluir pelo menos 'O' 'n' e '\n'
                if (texto[0] == 'O' or texto[0] == 'o') and len(texto) >= 3:
                    if texto[1] == 'n' or texto[1] == 'N':
                        # detetado "on"
                        somar = 1
                        texto = texto[2:] #ignorar caracteres lidos
                    else:
                        texto = texto[1:] #ignorar caracteres lidos  
                else:
                    # Neste caso, não vamos procurar por números
                    texto = texto[1:] #ignorar caracteres lidos

            else: # somar == 1 -> comportamento ligado
                if texto[0] == 'O' or texto[0] == 'o':
                    if texto[1] == 'f' or texto[1] == 'F':
                        if texto[2] == 'f' or texto[2] == 'F':
                            #Detetado "off"
                            somar = 0
                            texto = texto[3:] #ignorar caracteres lidos
                        else:
                            texto = texto[2:] #ignorar caracteres lidos
                    else:
                        texto = texto[1:] #ignorar caracteres lidos
                else:
                    fracionario = 0 # ATENÇÃO: só pode estar um ponto "." ou "," para ser um número fracionário
                    if texto[0].isnumeric(): # 1º caracter é um número (temos de ver até onde termina)
                        fracionario = 0 #por enquanto, é um número com casas decimais
                        e_numero = 1 # por enquanto, é um número
                        s_numero = ""
                        s_numero += texto[0]
                        texto = texto[1:]
                        while e_numero and (texto != ""):
                            if texto[0].isnumeric(): # verifica-se que é um número
                                s_numero += texto[0]
                                texto = texto[1:]
                            # temos de confirmar se tem casas decimais
                            elif not fracionario and (texto[0] == '.' or texto[0] == ','):
                                texto = texto[1:]
                                # mas ainda é preciso garantir que no espaço seguinte haja números
                                # senão, já não é considerado número
                                if texto == "": # número inválido
                                    e_numero = 0
                                elif texto[0].isnumeric():
                                    fracionario = 1
                                    s_numero += '.'
                                    s_numero += texto[0]
                                    texto = texto[1:]
                                else:
                                    e_numero = 0
                            else: # já não tem mais números OU encontramos uma virgula/ponto a mais
                                e_numero = 0
                        if fracionario: numero = float(s_numero) # ignora-se o elem de indice i
                        else: numero = int(s_numero)
                        if negativo: soma_total -= numero
                        else: soma_total += numero
                        numero = 0 # volta ao normal
                    # Caso não inicie por um número
                    else:
                        texto = texto[1:] # elimina caracter lido

if __name__ == '__main__':
    main()

Escreva uma linha: 22.73rt gb b7 7c  =364rbOff4tc 872cOff dshond672brd -266=21Ondcfvbnh2j2h e2 2h 5
Soma: 36.730000000000004
Soma: 806.73


KeyboardInterrupt: 

__Linha exemplo para teste__: 22.73rt gb b7 7c  =364rbOff4tc 872cOff dshond672brd -266=21Ondcfvbnh2j2h e2 2h 5

_Sequência de strings válidas_: 22.73, 7, 7, (=), 364, "Off" 4, 872, 672, -266, (=), 210, "On", 2, 2, 2, 2, 5

_Resultado esperado_:
Soma: 36.73
Soma: 806.73